pkg/controllers/generic/syncer.go

```diff
@@ -0,0 +1,310 @@
+			// this happens when some field is being removed shortly after being added, which suggest it's a timing issue
+			// it doesn't seem to have any negative consequence besides the logged error message
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		f.EventRecorder().Eventf(vObj, "Warning", "SyncError", "Error syncing to physical cluster: %v", err)
+		return ctrl.Result{}, fmt.Errorf("error applying patches: %v", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+var _ syncer.UpSyncer = &exporter{}
+
+func (f *exporter) SyncUp(ctx *synccontext.SyncContext, pObj client.Object) (ctrl.Result, error) {
+	if !translate.IsManaged(pObj) || f.isExcluded(pObj) {
+		return ctrl.Result{}, nil
+	}
+
+	// delete physical object because virtual one is missing
+	return syncer.DeleteObject(ctx, pObj, fmt.Sprintf("delete physical %s because virtual is missing", pObj.GetName()))
+}
+
+func (f *exporter) getControllerID() string {
+	if f.config.ID != "" {
+		return f.config.ID
+	}
+	// return plugin.GetPluginName()
+	return strings.Join(append(strings.Split(f.config.APIVersion, "/"), f.config.Kind), "-")
+}
+
+// TranslateMetadata converts the virtual object into a physical object
+func (f *exporter) TranslateMetadata(vObj client.Object) client.Object {
+	pObj := f.NamespacedTranslator.TranslateMetadata(vObj)
+	labels := pObj.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[controlledByLabel] = f.getControllerID()
+	pObj.SetLabels(labels)
+	return pObj
+}
+
+func (f *exporter) IsManaged(pObj client.Object) (bool, error) {
+	if !translate.IsManaged(pObj) {
+		return false, nil
+	}
+
+	return !f.isExcluded(pObj), nil
+}
+
+func isControlled(obj client.Object) bool {
+	return obj.GetLabels() != nil && obj.GetLabels()[controlledByLabel] != ""
+}
+
+func (f *exporter) objectMatches(obj client.Object) bool {
+	return f.selector == nil || !f.selector.Matches(labels.Set(obj.GetLabels()))
+}
+
+type virtualToHostNameResolver struct {
+	namespace       string
+	targetNamespace string
+}
+
+func (r *virtualToHostNameResolver) TranslateName(name string, regex *regexp.Regexp, _ string) (string, error) {
+	return r.TranslateNameWithNamespace(name, r.namespace, regex, "")
+}
+
+func (r *virtualToHostNameResolver) TranslateNameWithNamespace(name string, namespace string, regex *regexp.Regexp, _ string) (string, error) {
+	if regex != nil {
+		return patchesregex.ProcessRegex(regex, name, func(name, ns string) types.NamespacedName {
+			// if the regex match doesn't contain namespace - use the namespace set in this resolver
+			if ns == "" {
+				ns = namespace
+			}
+
+			translate.Default.PhysicalNamespace(namespace)
+			return types.NamespacedName{Namespace: r.targetNamespace, Name: translate.Default.PhysicalName(name, ns)}
```

Comment:
This part seems wrong. For one, `translate.Default.PhysicalNamespace(namespace)` returns a string, but it is never captured and used.

Reply:
yea nice catch, I've changed this to 
```go
return types.NamespacedName{
  Namespace: translate.Default.PhysicalNamespace(namespace),
  Name:      translate.Default.PhysicalName(name, ns)}
```